src/main-process/atom-window.coffee

```diff
@@ -128,8 +128,12 @@ class AtomWindow
         false
 
   handleEvents: ->
-    @browserWindow.on 'close', =>
-      @atomApplication.saveState(false)
+    @browserWindow.on 'close', (event) =>
+      unless @atomApplication.quitting or @unloading
+        event.preventDefault()
+        @unloading = true
+        @atomApplication.saveState(false)
+        @saveState().then(=> @close())
```

Comment:
> I could see a weird case where close is fired twice before the state finishes saving.

Yeah, depending on how big the state is (and, thus, how long it takes to serialize it), the user might be able to click the exit button more than once in rapid succession and cause the state to be serialized twice. I don't think it would be extremely, but it seems better to avoid it if we can.